Add hessian to _trustregion.py callback

SebastianSemper opened the pull request:
fixes #19719

SebastianSemper committed:
Add hessian to _trustregion.py callback

fixes #19719

mdhaber commented:
To prepare this PR for review, please:

- add the gradient, too, so we don't skip straight to Hessian,
- do this for all methods of `minimize` that support it, not just `trust-constr`,
- check that the values are present and as expected in the [existing test] (https://github.com/scipy/scipy/blob/309bfd5c257737cbf44ff1d6a0ac284db0c6f709/scipy/optimize/tests/test_optimize.py#L1602), and
- mention it in the [documentation](https://github.com/scipy/scipy/blob/5e4a5e3785f79dd4e8930eed883da89958860db2/scipy/optimize/_minimize.py#L208).

SebastianSemper commented:
I have tried to find other minimization routines, but it seems I am not familiar enough with the scipy codebase. I have trouble finding the code for these functions:

![image](https://github.com/scipy/scipy/assets/1104198/f47e6ea5-7993-4587-9007-0484467100b8)

Taking a look at the `_optimize.py`-file I could not find any definition of the supposedly imported functions. So I could need some help on that.

Other than that I have now added the jacobian to `intermediate_results`, but so far only for the trust-region methods. If I understood it correctly the different trust-region methods only differ in how they approach the trust-region subproblem...

SebastianSemper committed:
Add jacobian to intermediate_results in trustregion

mdhaber commented:
The functions are defined in `_optimize.py`. For example:

https://github.com/scipy/scipy/blob/e72de0714f0ad6e9b93c9107be84ee68ef70a5db/scipy/optimize/_optimize.py#L2054

Forms the `intermediate_result` here:

https://github.com/scipy/scipy/blob/e72de0714f0ad6e9b93c9107be84ee68ef70a5db/scipy/optimize/_optimize.py#L2230

SebastianSemper committed:
Add jac and hess to callback in more functions

 - added `jac` and `hess` to newton and bfgs methods
 - adopted the docstring accordingly
 - added value checks in the tests of the callback function

SebastianSemper commented:
I only looked at `__all__` in `_optimize.py` and ignored that the files is thousands of lines long... My bad.

I hope I have addressed your suggestions well enough...

mdhaber commented:
Ah, `__all__` is only used for the public methods.

mdhaber commented on scipy/optimize/_minimize.py:
```diff
@@ -206,10 +206,12 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
-        the callback raises ``StopIteration``.
+        of the parameter vector and objective function. If `jac` or `hess` are
+        specified as `Callable` they will also be present in 
+        ``intermdiate_result`` as attributes ``jac`` and ``hess``.
```
```suggestion
        ``intermediate_result`` as attributes ``jac`` and ``hess``.
```

mdhaber commented on scipy/optimize/_minimize.py:
```diff
@@ -206,10 +206,12 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
-        to be passed an `OptimizeResult`. These methods will also terminate if
-        the callback raises ``StopIteration``.
+        of the parameter vector and objective function. If `jac` or `hess` are
+        specified as `Callable` they will also be present in 
```
```suggestion
        callable, they will also be present in 
```
Is `Callable` a class? I've only seen it used as English.

Also, often these functions estimate the gradient/Hessian numerically if they are not provided as callables. Are these estimates not available?

SebastianSemper replied:
`Callable` should be at leat something one can use as a type hint, so I thought I use it as such.

These estimates might be available, but I checked how the final results are usually returned and there for instance the hessian is only provided, if a function to evaluate the hessian analytically is given.

mdhaber replied:
> should be at least something one can use as a type hint, so I thought I use it as such.

If you can find precedent elsewhere in `scipy.optimize`, sounds good. Since it's not a parameter, it would be need to be in double backticks unless Sphinx recognizes it. Otherwise, please change it.

SebastianSemper replied:
I have changed it back to just callable.

mdhaber commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -1600,11 +1600,19 @@ def h(x):
         if new_cb_interface == 1:
             def callback_interface(*, intermediate_result):
                 assert intermediate_result.fun == f(intermediate_result.x)
+                if "jac" in intermediate_result.keys():
```
I thought we had lists of the methods that support `jac` and `hess` defined somewhere like 

https://github.com/scipy/scipy/blob/46ce4bff44d1ebfe0bf1e9ea39c88b6be34defb1/scipy/optimize/_minimize.py#L40

Perhaps they are only defined somewhere in the test file. I can't look thoroughly right now, but I'd prefer to use such a list to determine whether to check the assertion because we'd also want the assertion to fall if it's missing an attribute it should have.

SebastianSemper replied:
`nelder-mead` for instance does not evaluate any gradients, neither do `powel` and `cg`. It took me some time to figure out that fprime, which is popping up at some places, is basically LaTeX-code. Afterwards I stepped through `_optimize.py` multiple times to find any jacobians and hessians.

I think one can also not define it on a per method basis, because in some cases the algorithms only approximate jacobians and hessians, or don't use them at all. Hence, the equality check you propose cannot be conducted purely based on the function tested alone since it depends on how `minimize` is called...

mdhaber replied:
In this test, callable gradients and Hessians are always provided.

SebastianSemper replied:
I added `MINIMIZE_METHODS_DERIV` and put the respective methods on there and use this list in the test.

mdhaber reviewed (commented):
At a glance it looks pretty complete. Later I'll need to check that this covers all the methods it should, but here are some preliminary thoughts.

SebastianSemper committed:
Fix type in docstring

SebastianSemper committed:
Refactor testing of callback

In order to have a more robust decision whether jacobians and gradients should have been passed to the callback, we add a new list of methods that do it and decide in the test, if we should test this.

mdhaber commented on scipy/optimize/_minimize.py:
```diff
@@ -41,6 +41,10 @@
                     'dogleg', 'trust-ncg', 'trust-exact', 'trust-krylov']
 
+# These methods may possibly pass jacobians or hessians to the callback
+MINIMIZE_METHODS_DERIV = ['bfgs', 'newton-cg', 'trust-constr',
+                        'dogleg', 'trust-ncg', 'trust-exact', 'trust-krylov']
```
According to the `minimize` documentation, the list of methods that support `jac` is different from those that support `hess`. If this list is for `jac`, it also needs `'cg', 'l-bfgs-b', 'tnc', 'slsqp'`. Adding this should reveal that `cg` and `l-bfgs-b` should be including `jac` in `intermediate_result` but currently aren't. (`tnc` and `slsqp` are exempt because they do not support the `intermediate_result` version of the callback interface.)

The list says that `hess` is "Only for Newton-CG, dogleg, trust-ncg, trust-krylov, trust-exact and trust-constr."

```suggestion
MINIMIZE_METHODS_JAC = ['cg', 'bfgs', 'newton-cg', 'l-bfgs-b', 'tnc', 'slsqp',
                        'dogleg', 'trust-ncg',  'trust-krylov', 'trust-exact', 
                        'trust-constr']
MINIMIZE_METHODS_HESS = ['newton-cg', 'dogleg', 'trust-ncg',  'trust-krylov',
                         'trust-exact', 'trust-constr']
```

mdhaber commented on scipy/optimize/_minimize.py:
```diff
@@ -206,10 +210,12 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
-        of the parameter must be ``intermediate_result`` for the callback
-        to be passed an `OptimizeResult`. These methods will also terminate if
-        the callback raises ``StopIteration``.
+        of the parameter vector and objective function. If ``jac`` or ``hess`` 
+        are specified as callable they will also be present in 
+        ``intermediate_result`` as attributes ``jac`` and ``hess``.
```
You had written:

> These estimates might be available, but I checked how the final results are usually returned and there for instance the hessian is only provided, if a function to evaluate the hessian analytically is given.

but the following methods report a `jac` estimate even if they are not provided a callable: `'cg', 'bfgs', 'l-bfgs-b', 'tnc', 'slsqp'`

<details>

```python3
from scipy import optimize
methods = ['cg', 'bfgs', 'l-bfgs-b', 'tnc', 'slsqp', 'trust-constr']
for method in methods:
    res = optimize.minimize(optimize.rosen, [1, 2, 3], method=method)
    print(f"{method}: {getattr(res, 'jac', None)}")
```
`jac` reported by ` trust-constr` is an empty list, which is a different issue...

</details>

`tnc` and `slsqp` don't have the `intermediate_result` callable interface, but I'd expect `'cg', 'bfgs', 'l-bfgs-b'` to have `jac` in the `intermediate_result` even if the user doesn't provide a callable.

But you're right, the other methods don't seem to report `hess` unless `hess` is a callable.

```suggestion
        of the parameter vector and objective function. If ``jac`` or ``hess`` 
        are available in the returned result object, they will also be present in
        ``intermediate_result`` as attributes ``jac`` and ``hess``.
```

mdhaber commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -23,6 +23,7 @@
 from scipy.optimize._minimize import Bounds, NonlinearConstraint
 from scipy.optimize._minimize import (MINIMIZE_METHODS,
                                       MINIMIZE_METHODS_NEW_CB,
+                                      MINIMIZE_METHODS_DERIV,
```
```suggestion
                                      MINIMIZE_METHODS_JAC,
                                      MINIMIZE_METHODS_HESS,
```

SebastianSemper replied:
Agreed.

mdhaber commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -1600,11 +1601,17 @@ def h(x):
         if new_cb_interface == 1:
             def callback_interface(*, intermediate_result):
                 assert intermediate_result.fun == f(intermediate_result.x)
+                if method in MINIMIZE_METHODS_DERIV:
+                    assert intermediate_result.jac == g(intermediate_result.x)
+                    assert intermediate_result.hess == h(intermediate_result.x)
```
These are arrays, right? 

```suggestion
                if method in MINIMIZE_METHODS_JAC:
                    assert_allclose(intermediate_result.jac, g(intermediate_result.x))
                if method in MINIMIZE_METHODS_HESS:
                    assert_allclose(intermediate_result.hess, h(intermediate_result.x))
```

SebastianSemper replied:
Sure. I must have been on something when writing this...

mdhaber commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -1600,11 +1601,17 @@ def h(x):
             class Callback:
                 def __call__(self, intermediate_result: OptimizeResult):
                     assert intermediate_result.fun == f(intermediate_result.x)
+                    if method in MINIMIZE_METHODS_DERIV:
+                        assert intermediate_result.jac == g(intermediate_result.x)
+                        assert intermediate_result.hess == h(intermediate_result.x)
```
```suggestion
                if method in MINIMIZE_METHODS_JAC:
                    assert_allclose(intermediate_result.jac, g(intermediate_result.x))
                if method in MINIMIZE_METHODS_HESS:
                    assert_allclose(intermediate_result.hess, h(intermediate_result.x))
```

mdhaber reviewed (changes requested):
Looks pretty close. The main thing is that methods CG and L-BFGS-B need to have their jacobian estimate added to `intermediate_result`.

Ideally, we'd also test what happens when `jac` is *not* provided as a callable for `'cg', 'bfgs', and 'l-bfgs-b'`. This could be added at the bottom of the test: if the method is one of these, call `minimize` again without passing `jac` or `hess`. The assertions in the callback should still pass, albeit maybe with a looser tolerance in `assert_allclose` because the estimated jacobian might not be as accurate as the callable jacobian.

Looks like many tests are failing. Please make sure tests pass before pushing.

SebastianSemper commented:
Alright, first I need to apologize for my former commits, which were not really good and not up to my usual standard. Sorry for wasting your time and I highly appreciate your patience with me.

1. I added the hessian as a return value to `newtoncg` as it was missing and also to the result fed into the callback.
2. I disagree with the necessity for calling minimize without `g` and `h`, since all this would test is the quality of the gradient approximation, which depends on the local Lipschitz-behavior of `f`.
3. I have adapted the docstring to account for the fact that now jacobians are always provided, no matter if approximated or not. Hessians are only provided, if present as callable.
4. Tests are now passing locally, I hope it stays the same on the remote..

SebastianSemper committed:
Fixes behavior of minimize callback and its tests

Adds jacobian to (hopefully) all methods that use it.
Fixes the test to use correct equality tests
Adapts docstring accordingly

mdhaber commented:
1. It looks like it wasn't added to `MINIMIZE_METHODS_HESS`, though?
2. 
> since all this would test is the quality of the gradient approximation

That wasn't the intent. I don't know the internals of all these functions, and it's conceivable that they could use a Jacobian if it is provided and not use a Jacobian if it's not provided. For example, `newton` (a scalar root finder) defaulted to the secant method if no derivative was provided, so it would not estimate the derivative on its own. We want to know which is the case to provide useful documentation that avoids bug reports later on.

SebastianSemper commented on scipy/optimize/_optimize.py:
```diff
@@ -2130,6 +2130,8 @@ def terminate(warnflag, msg):
                                 njev=sf.ngev, nhev=hcalls, status=warnflag,
                                 success=(warnflag == 0), message=msg, x=xk,
                                 nit=k)
+        if fhess is not None:
```
5. I don't know.
6. I won't. I am just closing this PR. Thank you.

melissawm replied:
Actually, what I don't understand here is the attribution - it looks like A is not defined here. It might be that we want the other way around (`A = result.hess`?) , but from skimming the code I think we may not need these two lines (tests pass withouth them)

SebastianSemper commented on scipy/optimize/_minimize.py:
```diff
@@ -205,11 +211,16 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
+        `OptimizeResult` with attributes ``x`` and ``fun`` and ``jac``, the
+        present values of the parameter vector, the objective function and
+        its gradient. 
+        Note that ``jac`` is only present if the method is in some form
+        using gradients during iteration. 
```
I think if the user wants to access the `jac` attribute in the callback the question whether the user knows about the functions' internals or not is redundant.

melissawm replied:
My suggestion is to say "See each method's documentation for details", but indeed I think quite a few of those could use a documentation refresh. I'm not sure how we could go about this here unless we make a list of methods.

mdhaber commented:
@j-bowhay do you have a moment to help bring this over the finish line? I can make the additions to the tests; I think we'd just need to work out how to document the methods/conditions under which these new attributes are available.

j-bowhay commented:
I am quite busy preparing for university exams (unfortunately!) and have a few open prs that need some attention however I could probably get to this by the end of Jan.

mdhaber commented:
Understood @j-bowhay! Good luck with exams!

I'd like to try to finish it while it's still fresh in my mind to avoid starting from scratch mentally. @andyfaff @lucascolley do you have any bandwidth for this?

lucascolley commented:
> @lucascolley do you have any bandwidth for this?

Unfortunately, I'm in a similar position to Jake (my bandwidth for new PRs is going to be very low from now until summer really - need to catch up on some work I've neglected this past month or so).

andyfaff commented:
Sorry, I'm out until 9th Jan onwards

melissawm commented:
@mdhaber I'd love to take a look once I'm back from my break on Jan 8. Is that reasonable? I may need some help getting into this code but I've been meaning to for a long time, so it might be a good opportunity. Cheers!

mdhaber commented:
Sounds good @melissawm! Thanks, all.

mdhaber commented:
Hi @melissawm I went ahead and re-opened this, and I think you can just push to the existing branch.

mdhaber commented:
@melissawm were you still interested or shall I check in to see if the others have time now?

lucascolley commented:
@mdhaber I don't suppose this is still fresh in your mind, and it seems like there are lots of other things you are working on at the minute, but FYI I have finished exams and could pick this (or other things) up.

mdhaber commented:
Congrats on finishing exams! 
If you're particularly interested in this, I will continue to support it for your sake. But no, it's no longer fresh, so at this point I would just as soon let someone else take over so I don't have to redo the review work I've already done to catch up.
If you're stil interested in the array API conversions, you know where the list is : ) I'd be happy to continue tag-teaming the writing and reviewing.
